go/private/go_repository.bzl

```diff
@@ -25,67 +25,84 @@ def _go_repository_impl(ctx):
   else:
     fail("neither commit or tag is specified", "commit")
 
-  # TODO(yugui): support submodule?
-  # c.f. https://www.bazel.io/versions/master/docs/be/workspace.html#git_repository.init_submodules
-  remote = ctx.attr.remote
-  vcs = ctx.attr.vcs
-  importpath = ctx.attr.importpath
-  result = ctx.execute([
-      fetch_repo,
-      '--dest', ctx.path(''),
-      '--remote', remote,
-      '--rev', rev,
-      '--vcs', vcs,
-      '--importpath', importpath])
-  if result.return_code:
-    fail("failed to fetch %s: %s" % (ctx.name, result.stderr))
+  if ctx.attr.url:
+    # explicit source url
+    if ctx.attr.vcs:
+        fail("cannot specify both of vcs and url")
+    ctx.download_and_extract(
+        url = ctx.attr.url,
+        sha256 = ctx.attr.sha256,
+        stripPrefix = ctx.attr.strip_prefix,
+        type = ctx.attr.type,
+    )
+  else:
+    # Using fetch repo
+    if ctx.attr.vcs and not ctx.attr.remote:
+      fail("if vcs is specified, remote must also be")
+    # TODO(yugui): support submodule?
+    # c.f. https://www.bazel.io/versions/master/docs/be/workspace.html#git_repository.init_submodules
+    result = ctx.execute([
+        ctx.path(ctx.attr._fetch_repo),
+        '--dest', ctx.path(''),
+        '--remote', ctx.attr.remote,
+        '--rev', rev,
+        '--vcs', ctx.attr.vcs,
+        '--importpath', ctx.attr.importpath,
+    ])
+    if result.return_code:
+      fail("failed to fetch %s: %s" % (ctx.name, result.stderr))
 
+  if (ctx.attr.disable_build_file_generation or 
+      ctx.path('BUILD').exists or 
```

Comment:
We don't support repo's without a root BUILD that sets the go_prefix...
But assuming we start to, either gazelle will run, add a root build file, and not damage the sub directory ones (which hopefully have all been generated by people using gazelle anyway) or you disable it with disable_build_file_generation.